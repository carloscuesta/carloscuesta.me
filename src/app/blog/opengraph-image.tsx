import { ImageResponse } from 'next/og'
import OpengraphImage, { getFonts } from 'src/components/OpengraphImage'

export const dynamic = 'force-static'
export const alt = 'Carlos Cuesta'
export const size = {
  width: 1200,
  height: 630,
}

export default async function Image() {
  return new ImageResponse(
    (
      <OpengraphImage
        title="Blog"
        url="blog"
        description="Explore my collection of writings and thoughts"
      />
    ),
    {
      fonts: await getFonts(),
    },
  )
}
