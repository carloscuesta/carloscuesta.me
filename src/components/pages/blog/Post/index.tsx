import Link from 'next/link'

import { type PostPreview } from 'src/utils/api/blog/mutators'

const Post = ({ post }: { post: PostPreview }) => (
  <li key={post.slug} className='rounded-lg font-semibold transition-all hover:bg-neutral-100 focus-within:bg-neutral-100 dark:hover:bg-neutral-900 dark:focus-within:bg-neutral-900 -mx-3'>
    <Link
      href={`/blog/${post.slug}`}
      title={post.title}
      className='rounded-lg flex flex-col-reverse gap-1 sm:flex-row sm:gap-3 px-3 py-3 sm:py-4 outline-none focus:ring-2 focus:ring-neutral-300 dark:focus:ring-neutral-500'
    > 
      <time
        className='opacity-60 sm:w-20'
        dateTime={post.datePublished.value}
        aria-label={`Written ${post.datePublished.formatDate}`}
      >
        {post.datePublished.formatMonthDay}
      </time>

      <p>{post.title}</p>
    </Link>
  </li>
)

export default Post
