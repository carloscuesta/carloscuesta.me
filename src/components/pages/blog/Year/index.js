// @flow
import type { Element } from 'react'

import styles from './styles.module.css'

type Props = {
  year: string
}

const Year = (props: Props): Element<'div'> => (
  <div className='col-xs-12'>
    <h2 className={styles.year}>{props.year}</h2>
  </div>
)

export default Year
