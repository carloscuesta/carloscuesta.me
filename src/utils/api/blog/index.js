// @flow
import fs from 'fs'
import { join } from 'path'
import matter from 'gray-matter'
import remark from 'remark'
import remarkHtml from 'remark-html'
import remarkExternalLinks from 'remark-external-links'

import { type Post, transformPost } from './mutators'

const POSTS_DIRECTORY: string = join(process.cwd(), 'src/posts')

export const getPostSlugs = (): Array<string> => fs.readdirSync(POSTS_DIRECTORY)
  .map((post: string) => post.replace('.md', ''))

export const fetchPost = async (slug: string): Promise<Post> => {
  const post = fs.readFileSync(join(POSTS_DIRECTORY, `${slug}.md`), 'utf8')
  const { data, content } = matter(post)
  const html = await remark()
    .use(remarkExternalLinks)
    .use(remarkHtml)
    .process(content)

  return transformPost({ data, html, slug })
}

export const fetchPosts = async (): Promise<Array<Post>> => {
  const posts: Array<Post> = await Promise.all(
    getPostSlugs().map((slug) => fetchPost(slug))
  )

  return posts.sort((x, y) => new Date(y.datePublished.value) - new Date(x.datePublished.value))
}
