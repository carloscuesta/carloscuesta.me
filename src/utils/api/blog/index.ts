import fs from 'fs'
import { join } from 'path'
import matter from 'gray-matter'
import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkGFM from 'remark-gfm'
import remarkToc from 'remark-toc'
import remarkRehype from 'remark-rehype'
import rehypeSlug from 'rehype-slug'
import rehypePrism from 'rehype-prism-plus'
import rehypeExternalLinks from 'rehype-external-links'
import rehypeAutoLinkHeadings from 'rehype-autolink-headings'
import rehypeStringify from 'rehype-stringify'
import rehypeMinify from 'rehype-preset-minify'
import rehypeWrap from 'rehype-wrap-all'

import { type Post, type PostPreview, transformPost } from './mutators'

const POSTS_DIRECTORY = join(process.cwd(), 'src/posts')

export const getPostSlugs = (): Array<string> => fs.readdirSync(POSTS_DIRECTORY)
  .map((post: string) => post.replace('.md', ''))

export const fetchPost = async (slug: string): Promise<Post> => {
  const post = fs.readFileSync(join(POSTS_DIRECTORY, `${slug}.md`), 'utf8')
  const { data, content } = matter(post)
  const html = await unified()
    .use(remarkParse)
    .use(remarkGFM)
    .use(remarkToc, { tight: true, maxDepth: 4 })
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeSlug)
    .use(
      rehypeAutoLinkHeadings,
      {
        content: {
          type: 'element',
          tagName: 'span',
          properties: { className: 'headingLink ' },
          children: [{ type: 'text', value: '#' }]
        }
      }
    )
    .use(rehypeExternalLinks)
    .use(rehypePrism, { ignoreMissing: true })
    .use(rehypeWrap, { selector:'table', wrapper: 'div.responsiveTable' })
    .use(rehypeStringify, { allowDangerousHtml: true })
    .use(rehypeMinify)
    .process(content)

  return transformPost({ data, html, slug })
}

export const fetchPosts = async (): Promise<Array<PostPreview>> => {
  const posts: Array<Post> = await Promise.all(
    getPostSlugs().map((slug) => fetchPost(slug))
  )

  return posts
    .sort((x, y) => {
      return Number(new Date(y.datePublished.value)) - Number(new Date(x.datePublished.value))
    })
    .map((post) => ({
      datePublished: post.datePublished,
      excerpt: post.excerpt,
      images: post.images,
      slug: post.slug,
      title: post.title
    }))
}
