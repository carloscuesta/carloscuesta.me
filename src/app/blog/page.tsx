import { type Metadata } from 'next'
import groupBy from 'lodash.groupby'

import { fetchPosts } from 'src/utils/api/blog'
import { fetchViews } from 'src/utils/api/blog/views'
import Wrapper from 'src/components/Wrapper'
import PageTitle from 'src/components/PageTitle'
import Year from './components/Year'
import Post from './components/Post'

const getData = async () => {
  const posts = await fetchPosts()
  const views = await fetchViews()

  return {
    posts: groupBy(posts, (post) =>
      new Date(post.datePublished.value).getFullYear()
    ),
    views,
  }
}

export const metadata: Metadata = {
  title: 'Blog',
  alternates: {
    canonical: '/blog',
  },
}

const Blog = async () => {
  const { posts, views } = await getData()

  return (
    <Wrapper>
      <main>
        <PageTitle title="Blog" />
        <div className="grid gap-8">
          {Object.keys(posts)
            .sort(() => -1)
            .map((year) => (
              <section key={year}>
                <Year year={year} />

                <ul className="grid grid-flow-row gap-3">
                  {posts[year].map((post) => (
                    <Post
                      key={post.slug}
                      post={post}
                      views={views[post.slug]}
                    />
                  ))}
                </ul>
              </section>
            ))}
        </div>
      </main>
    </Wrapper>
  )
}

export default Blog
