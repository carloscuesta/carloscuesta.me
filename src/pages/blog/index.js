// @flow
import React from 'react'
import groupBy from 'lodash.groupby'
import { NextSeo } from 'next-seo'

import { fetchPosts } from 'src/utils/api/blog'
import { type Post } from 'src/utils/api/blog/mutators'
import BlogPost from 'src/components/shared/BlogPost'
import Wrapper from 'src/components/shared/Wrapper'

type Props = {
  posts: { [key: string]: Array<Post> }
}

const Blog = (props: Props) => (
  <>
    <NextSeo
      canonical='https://carloscuesta.me/blog'
      title='Carlos Cuesta – Front End Engineer – Blog'
    />
    <main>
      <Wrapper>
        {Object.keys(props.posts).sort(() => -1).map((year: string) => (
          <div className='row' key={year}>
            <div className='col-xs-12'>
              <h3>{year}</h3>
            </div>

            {props.posts[year].map((post) => (
              <BlogPost post={post} key={post.slug} />
            ))}
          </div>
        ))}
      </Wrapper>
    </main>
  </>
)

export const getStaticProps = async (): Promise<{ props: Props }> => {
  const posts: Array<Post> = await fetchPosts()

  return {
    props: {
      posts: groupBy(
        posts,
        (post) => new Date(post.datePublished.value).getFullYear()
      )
    }
  }
}

export default Blog
