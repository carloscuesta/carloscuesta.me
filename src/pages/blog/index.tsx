import { GetStaticProps, InferGetStaticPropsType } from 'next'
import groupBy from 'lodash.groupby'
import { NextSeo } from 'next-seo'

import { fetchPosts } from 'src/utils/api/blog'
import { type PostPreview } from 'src/utils/api/blog/mutators'
import PageTitle from 'src/components/shared/PageTitle'
import BlogPost from 'src/components/shared/BlogPost'
import Wrapper from 'src/components/shared/Wrapper'
import Year from 'src/components/pages/blog/Year'

const Blog = (props: InferGetStaticPropsType<typeof getStaticProps>) => (
  <>
    <NextSeo
      canonical='https://carloscuesta.me/blog'
      title='Carlos Cuesta â€“ Blog'
    />
    <main>
      <Wrapper>
        <PageTitle title='Blog' />
        <>
          {Object.keys(props.posts).sort(() => -1).map((year) => (
            <div className='row' key={year}>
              <Year year={year} />

              {props.posts[year].map((post) => (
                <BlogPost post={post} key={post.slug} />
              ))}
            </div>
          ))}
        </>
      </Wrapper>
    </main>
  </>
)

export const getStaticProps: GetStaticProps<{
  posts: { [key: string]: PostPreview[] }
}> = async () => {
  const posts: Array<PostPreview> = await fetchPosts()

  return {
    props: {
      posts: groupBy(
        posts,
        (post) => new Date(post.datePublished.value).getFullYear()
      )
    }
  }
}

export default Blog
