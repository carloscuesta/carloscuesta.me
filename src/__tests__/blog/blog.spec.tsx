import renderer from 'react-test-renderer'

import Blog, { getStaticProps } from 'src/pages/blog'
import { fetchPosts } from 'src/utils/api/blog'
import * as stubs from './stubs'

jest.mock('next-seo', () => ({ NextSeo: 'NextSeo' }))
jest.mock('src/utils/api/blog')

/*
  I'm mocking the components since those are unit tested.
  So there's no point to test the rendering of them twice.
  It's more easier to manage little snapshots than big ones.
*/

jest.mock('src/components/shared/BlogPost', () => 'BlogPost')

/* Mock Date.now to avoid updating snapshots for formatDistanceToNow */
Date.now = jest.fn(() => 1593025862540)

describe('blog', () => {
  describe('page', () => {
    it('should match blog page', () => {
      const wrapper = renderer.create(<Blog {...stubs.props} />)

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('getStaticProps', () => {
    beforeAll(() => {
      (fetchPosts as jest.Mock).mockReturnValue(stubs.posts)
    })

    it('should return posts as props', async () => {
      const props = await getStaticProps({})

      expect(fetchPosts).toHaveBeenCalledWith()

      expect(props).toEqual({
        props: {
          posts: stubs.props.posts
        }
      })
    })
  })
})
