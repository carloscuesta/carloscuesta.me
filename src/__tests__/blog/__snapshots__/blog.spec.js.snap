// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blog page should match blog page 1`] = `
Array [
  <NextSeo
    canonical="https://carloscuesta.me/blog"
    title="Carlos Cuesta – Blog"
  />,
  <main>
    <div
      className="wrapper"
    >
      <h3
        className="title"
      >
        Blog
      </h3>
      <div
        className="row"
      >
        <div
          className="col-xs-12"
        >
          <h4
            className="year"
          >
            2019
          </h4>
        </div>
        <BlogPost
          post={
            Object {
              "dateModified": "2019-10-01 22:15",
              "datePublished": Object {
                "formatDate": "01 October 2019",
                "formatInWords": "9 months ago",
                "value": "2019-10-01 22:15",
              },
              "disqusIdentifier": "5bc9ff9949ac5006c0a84f00",
              "excerpt": "A guide to create JavaScript monorepos with Lerna and Yarn Workspaces. Explaining what is a monorepo, what are they useful for and how to create one with a code example",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/JavaScript-monorepos-with-Lerna-and-Yarn-Workspaces.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/JavaScript-monorepos-with-Lerna-and-Yarn-Workspaces.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/JavaScript-monorepos-with-Lerna-and-Yarn-Workspaces.png",
                },
              },
              "slug": "javascript-monorepos-lerna-yarn-workspaces",
              "title": "JavaScript monorepos with Lerna and Yarn Workspaces",
            }
          }
        />
        <BlogPost
          post={
            Object {
              "dateModified": "2019-09-12 17:00",
              "datePublished": Object {
                "formatDate": "12 September 2019",
                "formatInWords": "9 months ago",
                "value": "2019-09-12 17:00",
              },
              "disqusIdentifier": "5d78d0fcf942665cd6becd9a",
              "excerpt": "I'm going to explain why it's important and how you can use error boundaries in a React-Native application to improve error resiliency 👨‍💻",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/handling-react-native-errors-with-error-boundaries.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/handling-react-native-errors-with-error-boundaries.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/handling-react-native-errors-with-error-boundaries.png",
                },
              },
              "slug": "managing-react-native-crashes-with-error-boundaries",
              "title": "Managing React-Native crashes with Error Boundaries",
            }
          }
        />
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-12"
        >
          <h4
            className="year"
          >
            2018
          </h4>
        </div>
        <BlogPost
          post={
            Object {
              "dateModified": "2018-10-16 12:10",
              "datePublished": Object {
                "formatDate": "16 October 2018",
                "formatInWords": "over 1 year ago",
                "value": "2018-10-16 12:10",
              },
              "disqusIdentifier": "5b6c646126d36606d1805ab3",
              "excerpt": "Creating scalable React components using the folder pattern. A simple way to organize and structure React Components.",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/scalable-react-components.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/scalable-react-components.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/scalable-react-components.png",
                },
              },
              "slug": "scalable-react-components-architecture",
              "title": "Scalable React Components architecture",
            }
          }
        />
        <BlogPost
          post={
            Object {
              "dateModified": "2018-10-15 09:53",
              "datePublished": Object {
                "formatDate": "19 September 2018",
                "formatInWords": "almost 2 years ago",
                "value": "2018-09-19 09:53",
              },
              "disqusIdentifier": "5ae59dcdb3211a06522ad69b",
              "excerpt": "The process of continuously delivering React Native apps with Fastlane and Travis CI automatically. ",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/shipping-react-native-fastlane-travis.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/shipping-react-native-fastlane-travis.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/shipping-react-native-fastlane-travis.png",
                },
              },
              "slug": "shipping-react-native-fastlane-travis",
              "title": "Shipping React Native apps with Fastlane and Travis",
            }
          }
        />
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-12"
        >
          <h4
            className="year"
          >
            2017
          </h4>
        </div>
        <BlogPost
          post={
            Object {
              "dateModified": "2017-10-30 10:00",
              "datePublished": Object {
                "formatDate": "30 October 2017",
                "formatInWords": "over 2 years ago",
                "value": "2017-10-30 10:00",
              },
              "disqusIdentifier": "23",
              "excerpt": "ReactiveConf 2017 review. A two days conference about functional and reactive programming that takes places in Bratislava.",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1504460420/nmm99etv7j32h5lsulgx.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/v1504460420/nmm99etv7j32h5lsulgx.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1504460420/nmm99etv7j32h5lsulgx.png",
                },
              },
              "slug": "reactiveconf-2017",
              "title": "ReactiveConf 2017",
            }
          }
        />
        <BlogPost
          post={
            Object {
              "dateModified": "2017-10-18 10:15",
              "datePublished": Object {
                "formatDate": "18 October 2017",
                "formatInWords": "over 2 years ago",
                "value": "2017-10-18 10:15",
              },
              "disqusIdentifier": "59cfbaac613ac70679db193e",
              "excerpt": "The process of moving my website and Ghost blog from Heroku to DigitalOcean.  Provisioning up the server with Node.js, Nginx, LetsEncrypt and PM2.",
              "html": "<h1>html</h1>",
              "images": Object {
                "featured": Object {
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/moving-to-digitalocean.png",
                },
                "preview": Object {
                  "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_lqpi-post-preview/moving-to-digitalocean.png",
                  "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/moving-to-digitalocean.png",
                },
              },
              "slug": "moving-to-digitalocean",
              "title": "Moving to DigitalOcean",
            }
          }
        />
      </div>
    </div>
  </main>,
]
`;
